sycl/test/on-device/xocc/simple_tests/evaluation_tree.cpp

```diff
@@ -0,0 +1,235 @@
+// REQUIRES: xocc
+
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -fsycl-unnamed-lambda -std=c++20 -Xsycl-target-frontend -fno-exceptions %s -o %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+#include <CL/sycl.hpp>
+#include <cassert>
+#include <optional>
+#include <type_traits>
+#include <variant>
+namespace detail {
+
+template <typename> struct variant_trait {};
+
+template <typename... Tys> struct variant_trait<std::variant<Tys...>> {
+  using indexes = std::make_index_sequence<sizeof...(Tys)>;
+};
+
+// // This could be a replacement for visit_single_impl but causes addrespace casts
+// template <typename ret_type, typename Func, typename Var, auto... Idx>
+// inline ret_type
+// visit_single_impl(Func &&f, std::integer_sequence<size_t, Idx...>, Var &&var) {
+//   if constexpr (std::is_same<void, ret_type>::value) {
+//     (
+//         [&] {
+//           if (var.index() == Idx)
+//             std::forward<Func>(f)(std::get<Idx>(var));
+//         }(),
+//         ...);
+//   } else {
+//     // we use optional here because ret_type may not
+//     // be default constructible.
+//     std::optional<ret_type> ret;
+//     (
+//         [&] {
+//           if (var.index() == Idx)
+//             ret.emplace(std::forward<Func>(f)(std::get<Idx>(var)));
+//         }(),
+//         ...);
+//     return std::forward<ret_type>(*ret);
+//   }
+// }
+
+template <typename ret_type, typename Func, typename Var>
+[[noreturn]] inline ret_type
+visit_single_impl(Func&&, std::integer_sequence<size_t>, Var&&) {
+  // assert(false && "unreachable");
+  __builtin_unreachable();
+}
+
+template <typename ret_type, typename Func, typename Var, auto First,
+          auto... Idx>
+inline ret_type visit_single_impl(Func&& f,
+                                  std::integer_sequence<size_t, First, Idx...>,
+                                  Var&& var) {
+  if (var.index() == First)
+    return std::forward<Func>(f)(std::get<First>(var));
+  return visit_single_impl<ret_type>(std::forward<Func>(f),
+                                     std::integer_sequence<size_t, Idx...> {},
+                                     std::forward<Var>(var));
+}
+
+template <typename Func, typename Var>
+decltype(auto) visit_single(Func&& f, Var&& var) {
+  assert((!var.valueless_by_exception()));
+  using ret_type =
+      std::invoke_result_t<Func, decltype(std::get<0>(std::declval<Var>()))>;
+  return visit_single_impl<ret_type>(
+      std::forward<Func>(f),
+      typename variant_trait<
+          std::remove_cv_t<std::remove_reference_t<Var>>>::indexes {},
+      std::forward<Var>(var));
+}
+} // namespace detail
+
+/// dev_visit is std::visit implementation suitable to be used in device code.
+/// this version of visit doesn't use any function pointer but uses if series
```

Comment:
```suggestion
/// This version of visit doesn't use any function pointer but uses if series
```
I do not understand what you mean by "if series" and the following,